### PR TITLE
fix(icons): change tooling to build icons

### LINKED ISSUE #89: [Icons] The icons do not render in Firefox Nightly because the font resources are invalid
## Issue Type
Bug Report

## Component or Pattern Affected
The whole icon library

## Expected Behavior
I am expecting the icons to appear properly on screen

## Current Behavior
They all look like this:
screenshot 2019-02-26 at 10 15 15

## Possible Solution
Before mentioning possibles solutions, let's understand why this issue is happening.

After various tests, I concluded that the bug only happened in Firefox Nightly (not in Firefox beta nor stable) so [I filed a regression on bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1530624).

Two Firefox engineers got back to me explaining that Firefox **Nightly** is enforcing stricter parsing of webfonts on purpose, ie. to detect badly formed font resources. This stricter enforcement will not be activated in stable yet, but it may happen someday.

There is a [similar issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1492786) about another font.

Possible Solutions:
1. Consider it a non-issue since it only affects Firefox Nightly (users can set `gfx.downloadable_fonts.otl_validation` to `false` in their settings to ~~fix~~ hide this issue).
2. Investigate **why** the font generated is invalid, and fix it.
3. Use SVG for the icons because they are graphics and not fonts? Bring down the pitchforks, I am trolling 😅 . But genuinely, wasn't it possible to keep icons as SVG **and** make tree-shaking / dead code elimination work to avoid bundling the huge iconsData.json?

## Steps to Reproduce (for bugs)
1. Visit https://momentum.design/styles/icons/library with [Firefox Nightly](https://www.mozilla.org/en-US/firefox/channel/desktop/#nightly)

## Context
There are dozens of us... DOZENS (using Firefox Nightly as the daily browser)!

## Your Environment
Firefox Nightly, specifically.

### PR DESCRIPTION
**Pull request by tbassetto**
## Description

**This is a place do discuss the big changes introduced by this PR, it is not mergeable yet.**

* I replaced all the tooling but I did **not** commit the new generated files yet. I didn't want to add noise to the diff.
* Right now it depends on @tbassetto/webfonts-generator, until this module is deployed under @collab-ui/webfonts-generator.
* I kept handlebars as the templating solution to generate new files, but we no longer have to!
* I renamed all the files in `icons/templates` to include the full name of the generated file, I find it clearer.
* I deleted `icons/fonts/collab-ui-icons.eot` which I think is worth a breaking change I am afraid.
* `async`/`await` everywhere! Requires Node.JS 10+ I think.
* This new build scripts now detects svg files that can't be converted to `iconsData.json` and warns about them. There is currently 10 problematic icons or so.
* It's so much master to build icons now, I certainly hope we won't find any blockers!

Tomorrow I will test more thoroughly the generated fonts against know apps using the icons fonts.

## Related Issue
https://github.com/collab-ui/collab-ui/issues/89

## Motivation and Context
Wanted to fix #89.

## How Has This Been Tested?
Still working on it.

## Screenshots:
**Before** (If applicable):

**After**:

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## Checklist:
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: icons/tools/icons-build.js [hunk 1]
```diff
@@ -1,17 +1,61 @@
...
+    handlebars.registerHelper("equals", params => params[0] === params[1]);
+    handlebars.registerHelper("camelCase", str => camelCase(str));
+    handlebars.registerHelper("iosCharCode", str => "\\u{" + str + "}");
+
+    // Generate Scss files
+    const templateData = {
+      ...DEFAULT_TEMPLATE_OPTIONS,
+      ...result,
+      src: `url("#{$icon-font-path}/${path.parse(result.fontFiles.woff2).name}") format("woff2"),
+      url("#{$icon-font-path}/${path.parse(result.fontFiles.woff).name}") format("woff")`
```
Comment (tbassetto) [inline icons/tools/icons-build.js]: Note to myself: this code wrong, the file extensions are missing in the generated string. I'll fix it tomorrow.

File: icons/tools/generators.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        ds.push(d);
+      }
+    }
+    return { name, ds };
+  });
+
+  const iconsDs = await Promise.all(promises);
+  const icons = iconsDs.reduce((acc, curr) => {
+    acc[curr.name] = curr.ds;
+    if (curr.ds.length === 0) console.error(`⚠️ ${curr.name} has no paths!`);
```
Comment (tbassetto) [inline icons/tools/generators.js]: Current icons triggering this warning:
```
⚠️ blog_16 has no paths!
⚠️ content-share_20 has no paths!
⚠️ external-user_20 has no paths!
⚠️ people_18 has no paths!
⚠️ people-active_18 has no paths!
```

File: icons/tools/generators.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  await writeFile(finalFile, compile(data));
+  console.warn(`${finalFile} written!`);
+}
+
+async function generateIconsDataJson(data) {
+  const glyphs = data.glyphsData.map(({ file, name }) => ({
+    name,
+    file
+  }));
+  // const icons = {};
```
Comment (tbassetto) [inline icons/tools/generators.js]: Comment to be removed.

**Comment by tbassetto** [comment]
Fixed everything that I wanted to be fixed! The only thing missing is to actually depend on `@collab-ui/webfonts-generator` once pushed (though it can be done later actually!) and commit the new generated files (fonts, scss, css, json, swift, html).

It was much harder than I though to test the new font files. I managed to get them work on my local version of http://momentum.design but I only confirmed that they work on the sidebar. I couldn't load the icons from the icons page (because my local mongodb was empty?). At least icons in the sidebar looked exactly like in production **and** worked on Firefox Nightly.

**Comment by tbassetto** [comment]
I've rebased from master and committed the new fonts 😱.

You can see them on https://collab-ui-icons.now.sh/ (note that `adjust-video_16` does not look correct, but this icon is already funky on master).

Want to try the new font on a real website? Go to https://momentum.design/styles/icons/library (doesn't work on Atlas because of CSP) and type the following int the console:
```javascript
const link = document.createElement('link');
link.setAttribute('rel', 'stylesheet');
link.type = 'text/css';
link.href = 'https://collab-ui-icons.now.sh/collab-ui-icons.css';
document.head.appendChild(link);
```

It's a special stylesheet using the "new" font and adding a light gray outline to them (to be sure that you are using the new ones). Notice how everything is the same! Actually it's not, it's better because now the icons appear on Firefox Nightly.

**Comment by pauljeter** [comment]
Merged

**Pull request closed without merging**